File: solutions/python/22.py
from collections import deque


def can_escape(maze: list[list[str]]) -> bool:
    maze_length = len(maze)
    row_length = len(maze[0])
    directions = [
        [0, 1],
        [1, 0],
        [0, -1],
        [-1, 0],
    ]

    santa_position = [-1, -1]
    exit_position = [-1, -1]

    for i in range(0, maze_length):
        for j in range(0, row_length):
            if (maze[i][j] == "S"):
                santa_position = [i, j]
            if (maze[i][j] == "E"):
                exit_position = [i, j]

    queue = deque()
    queue.append([santa_position[0], santa_position[1]])
    visited = set()
    visited.add(f"{santa_position[0]},{santa_position[1]}")

    while len(queue) > 0:
        [x, y] = queue.popleft()

        if (x == exit_position[0] and y == exit_position[1]):
            return True

        for [dx, dy] in directions:
            new_x = x + dx
            new_y = y + dy

            if (
                new_x >= 0 and
                new_x < maze_length and
                new_y >= 0 and
                new_y < row_length and
                maze[new_x][new_y] != "#" and
                f"{new_x},{new_y}" not in visited
            ):
                visited.add(f"{new_x},{new_y}")
                queue.append([new_x, new_y])

    return False


escape1 = can_escape([
    ["S", ".", "#", "."],
    ["#", ".", "#", "."],
    [".", ".", ".", "."],
    ["#", "#", "#", "E"],


])
# → true
print(escape1)

escape2 = can_escape([
    ["S", "#", "#"],
    [".", "#", "."],
    [".", "#", "E"],
])
# → false
print(escape2)

escape3 = can_escape([["S", "E"]])
# → true
print(escape3)

escape4 = can_escape([
    ["S", ".", ".", ".", "."],
    ["#", "#", "#", "#", "."],
    [".", ".", ".", ".", "."],
    [".", "#", "#", "#", "#"],
    [".", ".", ".", ".", "E"],
])
# → true
print(escape4)

escape5 = can_escape([
    ["S", ".", "."],
    [".", ".", "."],
    ["#", "#", "#"],
    [".", ".", "E"],
])
# → false
print(escape5)


"""
  Score: ★★★★★★★★ (3 bonus stars for solving it today!)
  Points: 10
  ops: 2267
"""

File: solutions/js/22.js
/**
 * @param {string[][]} maze
 * @returns {boolean}
 */
function canEscape(maze) {
  const mazeLength = maze.length;
  const rowLength = maze[0].length;
  const directions = [
    [0, 1],
    [1, 0],
    [0, -1],
    [-1, 0],
  ];

  let santaPosition = [-1, -1];
  let exitPosition = [-1, -1];

  for (let i = 0; i < mazeLength; i++) {
    const santaCol = maze[i].indexOf("S");
    const exitCol = maze[i].indexOf("E");
    if (santaCol !== -1) santaPosition = [i, santaCol];
    if (exitCol !== -1) exitPosition = [i, exitCol];
  }

  const queue = [[santaPosition[0], santaPosition[1]]];
  const visited = new Set();
  visited.add(`${santaPosition[0]},${santaPosition[1]}`);

  while (queue.length > 0) {
    const [x, y] = queue.shift();

    if (x === exitPosition[0] && y === exitPosition[1]) {
      return true;
    }

    for (const [dx, dy] of directions) {
      const newX = x + dx;
      const newY = y + dy;

      if (
        newX >= 0 &&
        newX < mazeLength &&
        newY >= 0 &&
        newY < rowLength &&
        maze[newX][newY] !== "#" &&
        !visited.has(`${newX},${newY}`)
      ) {
        visited.add(`${newX},${newY}`);
        queue.push([newX, newY]);
      }
    }
  }

  return false;
}

const escape1 = canEscape([
  ["S", ".", "#", "."],
  ["#", ".", "#", "."],
  [".", ".", ".", "."],
  ["#", "#", "#", "E"],
]);
// → true
console.log(escape1);

const escape2 = canEscape([
  ["S", "#", "#"],
  [".", "#", "."],
  [".", "#", "E"],
]);
// → false
console.log(escape2);

const escape3 = canEscape([["S", "E"]]);
// → true
console.log(escape3);

const escape4 = canEscape([
  ["S", ".", ".", ".", "."],
  ["#", "#", "#", "#", "."],
  [".", ".", ".", ".", "."],
  [".", "#", "#", "#", "#"],
  [".", ".", ".", ".", "E"],
]);
// → true
console.log(escape4);

const escape5 = canEscape([
  ["S", ".", "."],
  [".", ".", "."],
  ["#", "#", "#"],
  [".", ".", "E"],
]);
// → false
console.log(escape5);

/* 
  Score: ★★★★★★★★ (3 bonus stars for solving it today!)
  Points: 10
  ops: 2177
*/
